src/abstractdataframe/selection.jl

```diff
@@ -0,0 +1,357 @@
+are supported. The syntax for column renaming is `old_column=>new_column_name`.
+The syntax for column transformations is `old_column=>fun=>new_column_name`.
+`old_column` can be a `Symbol` or an integer, ``new_column_name` must be a `Symbol`,
+and `fun` must be a `Function` that is applied row by row to the values of `old_column`.
+
+Column transfomation syntax also supports a short `old_column=>fun` form, in which
+case `new_column_name` is automatically generated by joining `olcd_column` with `fun` name
+with `_`.
+
+If more than one argument is passed then duplicates are accepted except for
+column renaming and transformation operations, where it is not allowed to rename/transform
+into the same column name.
+For example if `:col` is present in `df` a call `select(df, :col, :)` is valid
+and creates a new data frame with column `:col` moved to be the first.
+
+If `df` is a `DataFrame` a new `DataFrame` is returned.
+If `copycols=true` (the default), then returned `DataFrame` is guaranteed not to share columns with `df`.
+If `copycols=false`, then returned `DataFrame` shares column vectors with `df` where possible.
+
+If `df` is a `SubDataFrame` then a `SubDataFrame` is returned if `copycols=false`
+and a `DataFrame` with freshly allocated columns otherwise.
+
+Note that if `df` is a `DataFrame` and `copycols=false` then including the same column several times
+in the resulting data frame will create aliases.
+
+# Examples
+```jldoctest
+julia> df = DataFrame(a=1:3, b=4:6)
+3×2 DataFrame
+│ Row │ a     │ b     │
+│     │ Int64 │ Int64 │
+├─────┼───────┼───────┤
+│ 1   │ 1     │ 4     │
+│ 2   │ 2     │ 5     │
+│ 3   │ 3     │ 6     │
+
+julia> select(df, :b)
+3×1 DataFrame
+│ Row │ b     │
+│     │ Int64 │
+├─────┼───────┤
+│ 1   │ 4     │
+│ 2   │ 5     │
+│ 3   │ 6     │
+
+julia> select(df, Not(:b)) # drop column :b from df
+3×1 DataFrame
+│ Row │ a     │
+│     │ Int64 │
+├─────┼───────┤
+│ 1   │ 1     │
+│ 2   │ 2     │
+│ 3   │ 3     │
+
+julia> select(df, :a=>:c, :b)
+3×2 DataFrame
+│ Row │ c     │ b     │
+│     │ Int64 │ Int64 │
+├─────┼───────┼───────┤
+│ 1   │ 1     │ 4     │
+│ 2   │ 2     │ 5     │
+│ 3   │ 3     │ 6     │
+
+julia> select(df, :b, :a=><(1.5)=>:c)
+3×2 DataFrame
+│ Row │ b     │ c    │
+│     │ Int64 │ Bool │
+├─────┼───────┼──────┤
+│ 1   │ 4     │ 1    │
+│ 2   │ 5     │ 0    │
+│ 3   │ 6     │ 0    │
+```
+
+"""
+select(df::DataFrame, inds::AbstractVector{Int}; copycols::Bool=true) =
+    DataFrame(_columns(df)[inds], Index(_names(df)[inds]),
+              copycols=copycols)
+select(df::DataFrame, c::Union{AbstractVector{<:Integer}, AbstractVector{Symbol},
+                               Colon, All, Not, Between, Regex}; copycols::Bool=true) =
+    select(df, index(df)[c], copycols=copycols)
+select(df::DataFrame, c::ColumnIndex; copycols::Bool=true) =
+    select(df, [c], copycols=copycols)
+
+function select(df::DataFrame, cs...; copycols::Bool=true)
```

Comment:
To avoid recompiling the function for each combination of number and type of arguments, probably better have this method be a one-liner which calls `_select(df, collect(cs))` or `_select(df, [normalize_selection(index(df), c) for c in cs])` immediately.

Reply:
A good point. I have not thought about it. I will change.

Reply:
it has an additional benefit that this way `copycols` will be inlined.